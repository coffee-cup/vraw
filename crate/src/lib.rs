use wasm_bindgen::prelude::*;

mod error;
mod interpret;
mod lexer;
mod parser;
mod utils;

// When the `wee_alloc` feature is enabled, use `wee_alloc` as the global
// allocator.
#[cfg(feature = "wee_alloc")]
#[global_allocator]
static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;

#[wasm_bindgen]
pub struct Foo {
    test: String,
}

#[wasm_bindgen]
impl Foo {
    pub fn go_riders(&self, x: &str) -> String {
        let mut s = self.test.clone();
        s.push_str(x);
        s
    }
}

#[wasm_bindgen]
pub fn bar(x: &str) -> Foo {
    let foo = Foo {
        test: "hello world".to_owned(),
    };
    foo
}

#[wasm_bindgen]
pub fn compile(input: &str) -> String {
    let tokens = match lexer::lex(&input.to_owned()) {
        Ok(tokens) => tokens,
        Err(err) => return format!("{}", err),
    };

    let program = match parser::parse_program(tokens) {
        Ok(program) => program,
        Err(err) => return format!("{}", err),
    };

    let result = match interpret::eval_program(&program) {
        Ok(value) => value,
        Err(err) => {
            return format!("{}", err);
        }
    };

    result
}
