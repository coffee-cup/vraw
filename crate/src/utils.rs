#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Pos {
    pub line: u32,
    pub column: u32,
}

#[derive(Debug, PartialEq, Clone, Copy)]
pub struct Range {
    pub start: Pos,
    pub end: Pos,
}

pub trait HasPos {
    fn pos(&self) -> Pos;
}

pub fn create_pos(line: u32, column: u32) -> Pos {
    Pos {
        line: line,
        column: column,
    }
}

pub fn create_range(start: Pos, end: Pos) -> Range {
    Range {
        start: start,
        end: end,
    }
}

pub fn set_panic_hook() {
    // When the `console_error_panic_hook` feature is enabled, we can call the
    // `set_panic_hook` function at least once during initialization, and then
    // we will get better error messages if our code ever panics.
    //
    // For more details see
    // https://github.com/rustwasm/console_error_panic_hook#readme
    #[cfg(feature = "console_error_panic_hook")]
    console_error_panic_hook::set_once();
}
