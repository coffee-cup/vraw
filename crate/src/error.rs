use std::fmt;

use crate::utils::*;

#[derive(Debug, PartialEq)]
pub struct Error<T: fmt::Display> {
    pub error_type: T,
    pub pos: Pos,
}

impl<T: fmt::Display> Error<T> {
    pub fn new(error_type: T, pos: Pos) -> Error<T> {
        Error {
            error_type: error_type,
            pos: pos,
        }
    }
}

impl<T: fmt::Display> HasPos for Error<T> {
    fn pos(&self) -> Pos {
        self.pos
    }
}

impl<T: fmt::Display> fmt::Display for Error<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "[{}:{}]: {}",
            self.pos.line, self.pos.column, self.error_type
        )
    }
}
