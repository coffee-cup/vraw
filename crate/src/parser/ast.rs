use crate::utils::*;

pub type Ident = String;
pub type Arg = Ident;

#[derive(Debug, PartialEq, Clone)]
pub struct Program {
    pub decls: Vec<Decl>,
    pub end: Pos,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Decl {
    ShapeDecl(Shape),
}

#[derive(Debug, PartialEq, Clone)]
pub struct Shape {
    pub name: Ident,
    pub args: Vec<Arg>,
    pub block: Block,
    pub pos: Pos,
}

#[derive(Debug, PartialEq, Clone)]
pub struct Block {
    pub calls: Vec<FunCall>,
    pub range: Range,
}

#[derive(Debug, PartialEq, Clone)]
pub struct NamedArg {
    pub name: Ident,
    pub expr: Expr,
}

#[derive(Debug, PartialEq)]
pub enum Stmt {
    Expr(Expr, Pos),
}

#[derive(Debug, PartialEq, Clone)]
pub enum Expr {
    Name(Ident, Range),
    Literal(Literal, Range),
    Binary(Box<Expr>, BinOp, Box<Expr>, Pos),
    Unary(UnOp, Box<Expr>, Pos),
    Grouping(Box<Expr>),
}

#[derive(Debug, PartialEq, Clone)]
pub struct FunCall {
    pub ident: Ident,
    pub args: Vec<NamedArg>,
    pub range: Range,
}

#[derive(Debug, PartialEq, Clone)]
pub enum Literal {
    Number(f64),
    String(String),
}

#[derive(Debug, PartialEq, Clone)]
pub enum BinOp {
    Mul,
    Div,
    Add,
    Sub,
}

#[derive(Debug, PartialEq, Clone)]
pub enum UnOp {
    Neg,
}

pub fn prec(p: Precedence) -> u32 {
    p as u32
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Precedence {
    Conditional = 20,
    Sum = 30,
    Product = 40,
    Exponent = 50,
    Prefix = 60,
    Postfix = 70,
    Call = 80,
}

impl HasPos for Expr {
    fn pos(&self) -> Pos {
        match self {
            Expr::Name(_, range) => range.start,
            Expr::Literal(_, range) => range.start,
            Expr::Binary(_, _, _, p) => *p,
            Expr::Unary(_, _, p) => *p,
            Expr::Grouping(ref e) => e.pos(),
        }
    }
}

impl HasPos for Stmt {
    fn pos(&self) -> Pos {
        match self {
            Stmt::Expr(_, p) => *p,
        }
    }
}

impl HasPos for FunCall {
    fn pos(&self) -> Pos {
        self.range.start
    }
}

impl HasPos for Shape {
    fn pos(&self) -> Pos {
        self.pos
    }
}
