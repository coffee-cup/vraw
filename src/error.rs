use std::fmt;

use crate::utils::*;

#[derive(Debug, PartialEq)]
pub struct Error {
    message: String,
    pos: Option<Pos>,
}

impl Error {
    pub fn new(message: String, pos: Option<Pos>) -> Error {
        Error {
            message: message,
            pos: pos,
        }
    }
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self.pos {
            Some(p) => write!(f, "[{}:{}]: {}", p.line, p.column, self.message),
            _ => write!(f, "{}", self.message),
        }
    }
}
