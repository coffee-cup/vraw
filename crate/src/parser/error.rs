use std::fmt;

use crate::error::Error;
use crate::utils::*;

#[derive(Debug, PartialEq)]
pub enum ParseErrorType {
    UnExpectedEndOfInput,
    IdentiferCannotBeReservedWord(String),
    UnBalancedParen,
    Expected(String, Option<String>),
}

impl fmt::Display for ParseErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            ParseErrorType::UnExpectedEndOfInput => write!(f, "Unexpected end of input."),
            ParseErrorType::IdentiferCannotBeReservedWord(id) => {
                write!(f, "Identifier {} cannot be a reserved word.", id)
            }
            ParseErrorType::UnBalancedParen => write!(f, "Unbalanced paren."),
            ParseErrorType::Expected(expected, found) => match found {
                Some(found) => write!(f, "Expected {}. Found {}.", expected, found),
                None => write!(f, "Expected {}", expected),
            },
        }
    }
}

pub type ParseError = Error<ParseErrorType>;
