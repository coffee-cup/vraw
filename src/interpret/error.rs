use std::fmt;

use crate::error::Error;
use crate::utils::*;

#[derive(Debug, PartialEq, Clone)]
pub enum EvalErrorType {
    VariableNotDefined(String),
    ShapeNotDefined(String),
    TypeMismatch(String, String),
    SvgExpectsString(String),
    ShapeAlreadyDefined(String),
    NumArgs(String, usize, usize),
    InvalidArgName(String, String),
    MissingArgs(String, Vec<String>),
    MissingMain,
}

impl fmt::Display for EvalErrorType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            EvalErrorType::VariableNotDefined(name) => write!(f, "Variable `{}` not defined", name),
            EvalErrorType::ShapeNotDefined(name) => write!(f, "Shape `{}` not defined", name),
            EvalErrorType::TypeMismatch(expected, received) => write!(
                f,
                "Typemismatch. Expected: {}, Received: {}",
                expected, received
            ),
            EvalErrorType::SvgExpectsString(received) => write!(
                f,
                "svg value arg needs to be a string. Received {}",
                received
            ),
            EvalErrorType::ShapeAlreadyDefined(name) => write!(f, "Shape {} already defined", name),
            EvalErrorType::NumArgs(func, expected, received) => write!(
                f,
                "Incorrect number of args to {}. Expected: {}, Received: {}",
                func, expected, received
            ),
            EvalErrorType::InvalidArgName(func, arg) => {
                write!(f, "{} does not have an arg named {}", func, arg)
            }
            EvalErrorType::MissingArgs(func, args) => {
                write!(f, "Missing args {} for {}", args.join(", "), func)
            }
            EvalErrorType::MissingMain => write!(f, "Missing main shape"),
        }
    }
}

pub type EvalError = Error<EvalErrorType>;
