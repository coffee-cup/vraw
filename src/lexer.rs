use std::iter::Peekable;

use crate::error::*;
use crate::utils::*;

#[derive(Debug, PartialEq, Clone)]
pub enum TokenType {
    LParen,
    RParen,
    LCurly,
    RCurly,
    Times,
    Divide,
    Plus,
    Minus,
    Equals,
    Compare,
    Colon,
    Comma,
    Number(f64),
    Ident(String),
}

#[derive(Debug, PartialEq, Clone)]
pub struct Token {
    token_type: TokenType,
    token_pos: Range,
}

pub struct Lexer<'a> {
    iter: Peekable<std::str::Chars<'a>>,
    line: u32,
    column: u32,
}

pub fn lex(input: &String) -> Result<Vec<Token>, Error> {
    let mut tokens = vec![];
    let mut lexer = Lexer::new(input);

    while let Some(t) = lexer.next()? {
        tokens.push(t);
    }

    Ok(tokens)
}

impl Token {
    pub fn token_type(&self) -> TokenType {
        self.token_type.clone()
    }

    pub fn token_pos(&self) -> Range {
        self.token_pos
    }
}

impl<'a> Lexer<'a> {
    pub fn new(input: &'a String) -> Lexer<'a> {
        Lexer {
            iter: input.trim().chars().peekable(),
            line: 0,
            column: 0,
        }
    }

    fn pos(&self) -> Pos {
        Pos {
            line: self.line,
            column: self.column,
        }
    }

    fn token(&self, token_type: TokenType, start: Pos, end: Pos) -> Option<Token> {
        Some(Token {
            token_type: token_type,
            token_pos: Range {
                start: start,
                end: end,
            },
        })
    }

    fn advance(&mut self, token_type: TokenType) -> Result<Option<Token>, Error> {
        let start = self.pos();

        self.iter.next();
        self.column += 1;

        let end = self.pos();

        let token = self.token(token_type, start, end);
        Ok(token)
    }

    fn advance_whitespace(&mut self) -> bool {
        match self.iter.peek() {
            Some(' ') | Some('\t') => {
                self.column += 1;
                self.iter.next();
                true
            }
            Some('\n') | Some('\r') => {
                self.column = 0;
                self.line += 1;
                self.iter.next();
                true
            }
            _ => false,
        }
    }

    fn forward(&mut self) -> Option<char> {
        self.column += 1;
        self.iter.next()
    }

    fn match_next(&mut self, pred: &Fn(char) -> bool) -> Option<char> {
        let c = match self.iter.peek() {
            Some(c) => *c,
            None => return None,
        };

        if pred(c) {
            self.forward();
            Some(c)
        } else {
            None
        }
    }

    fn consume_ident(&mut self) -> Result<Option<Token>, Error> {
        let start = self.pos();
        let mut id = String::new();

        match self.match_next(&is_alpha) {
            Some(c) => id.push(c),
            None => {
                return Err(Error::new(
                    "identifiers must start with letter.".to_owned(),
                    Some(self.pos()),
                ))
            }
        };

        while let Some(c) = self.match_next(&is_alphanum) {
            id.push(c);
        }
        let end = self.pos();

        let token = self.token(TokenType::Ident(id), start, end);
        Ok(token)
    }

    fn consume_number(&mut self) -> Result<Option<Token>, Error> {
        let start = self.pos();

        let mut str = String::new();
        let mut seen_dot = false;

        while let Some(c) = self.iter.peek() {
            if *c == '.' && !seen_dot {
                seen_dot = true;
                str.push(self.forward().unwrap());
            } else if is_digit(*c) {
                str.push(self.forward().unwrap());
            } else {
                break;
            }
        }

        let end = self.pos();

        let n = str.parse::<f64>().unwrap();

        Ok(self.token(TokenType::Number(n), start, end))
    }

    fn next(&mut self) -> Result<Option<Token>, Error> {
        // ignore whitespace
        while self.advance_whitespace() {}

        let c = match self.iter.peek() {
            Some(c) => c,
            None => return Ok(None),
        };

        let token = match c {
            '(' => self.advance(TokenType::LParen),
            ')' => self.advance(TokenType::RParen),
            '{' => self.advance(TokenType::LCurly),
            '}' => self.advance(TokenType::RCurly),
            '*' => self.advance(TokenType::Times),
            '/' => self.advance(TokenType::Divide),
            '+' => self.advance(TokenType::Plus),
            '-' => self.advance(TokenType::Minus),
            ':' => self.advance(TokenType::Colon),
            ',' => self.advance(TokenType::Comma),
            '=' => {
                let start = self.pos();
                self.forward();
                let end = self.pos();
                match self.iter.peek() {
                    Some('=') => self.advance(TokenType::Compare),
                    _ => Ok(self.token(TokenType::Equals, start, end)),
                }
            }
            'a'...'z' => self.consume_ident(),
            '0'...'9' => self.consume_number(),
            _ => Err(Error::new(
                format!("unexpected character: {:?}", c),
                Some(self.pos()),
            )),
        };

        token
    }
}

fn is_alpha(c: char) -> bool {
    match c {
        'a'...'z' | 'A'...'Z' => true,
        _ => false,
    }
}

fn is_digit(c: char) -> bool {
    match c {
        '0'...'9' => true,
        _ => false,
    }
}

fn is_alphanum(c: char) -> bool {
    is_alpha(c) || is_digit(c)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lex_identifier() {
        let tokens = lex(&"a1234".to_owned());
        assert_eq!(
            tokens,
            Ok(vec![Token {
                token_type: TokenType::Ident("a1234".to_owned()),
                token_pos: Range {
                    start: Pos { line: 0, column: 0 },
                    end: Pos { line: 0, column: 5 }
                }
            }])
        );
    }

    #[test]
    fn lex_integer() {
        let tokens = lex(&"10".to_owned());
        assert_eq!(
            tokens,
            Ok(vec![Token {
                token_type: TokenType::Number(10.0),
                token_pos: Range {
                    start: Pos { line: 0, column: 0 },
                    end: Pos { line: 0, column: 2 }
                }
            }])
        )
    }

    #[test]
    fn lex_float() {
        let tokens = lex(&"10.123".to_owned());
        assert_eq!(
            tokens,
            Ok(vec![Token {
                token_type: TokenType::Number(10.123),
                token_pos: Range {
                    start: Pos { line: 0, column: 0 },
                    end: Pos { line: 0, column: 6 }
                }
            }])
        )
    }
}
