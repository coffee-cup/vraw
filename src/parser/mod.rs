use std::iter::Peekable;
use std::slice::Iter;

use crate::lexer::*;
use crate::utils::*;

pub mod ast;
mod error;

use ast::*;
use error::ParseErrorType::*;
use error::*;

const RESERVED: &'static [&'static str] = &["shape", "transform"];

struct Parser<'a> {
    input: Peekable<Iter<'a, Token>>,
    input_end_pos: Pos,
}

type ParseResult<T> = Result<T, ParseError>;

fn parse_error<T>(error_type: ParseErrorType, pos: Pos) -> ParseResult<T> {
    Err(ParseError {
        error_type: error_type,
        pos: pos,
    })
}

fn is_reserved_word(word: &str) -> bool {
    RESERVED.contains(&word)
}

pub fn parse_program(tokens: Vec<Token>) -> ParseResult<Program> {
    let mut parser = Parser::new(&tokens);
    parser.program()
}

pub fn parse_expression(tokens: Vec<Token>) -> ParseResult<Expr> {
    let mut parser = Parser::new(&tokens);
    parser.expression(0)
}

impl Token {
    // null denotation
    fn nud(&self, parser: &mut Parser) -> ParseResult<Expr> {
        match self.token_type() {
            TokenType::Ident(s) => {
                if is_reserved_word(s.as_str()) {
                    parse_error(IdentiferCannotBeReservedWord(s), self.token_pos().start)
                } else {
                    Ok(Expr::Name(s, self.token_pos()))
                }
            }
            TokenType::Number(n) => Ok(Expr::Literal(Literal::Number(n), self.token_pos())),
            TokenType::Minus => {
                let e = parser.expression(prec(Precedence::Prefix))?;
                Ok(Expr::Unary(UnOp::Neg, Box::new(e), self.token_pos().start))
            }
            TokenType::LParen => {
                let e = parser.expression(0)?;
                if let Some(TokenType::RParen) = parser.input.next().map(|t| t.token_type()) {
                    Ok(Expr::Grouping(Box::new(e)))
                } else {
                    parse_error(UnBalancedParen, self.token_pos().start)
                }
            }
            t => parse_error(
                Expected("literal".to_owned(), Some(format!("{:?}", t))),
                self.token_pos().start,
            ),
        }
    }

    // left denotation
    fn led(&self, parser: &mut Parser, lhs: Expr) -> ParseResult<Expr> {
        match self.token_type() {
            TokenType::Times | TokenType::Divide | TokenType::Plus | TokenType::Minus => {
                let rhs = parser.expression(self.lbp())?;
                let op = token_to_binop(self.clone()).unwrap();
                Ok(Expr::Binary(
                    Box::new(lhs),
                    op,
                    Box::new(rhs),
                    self.token_pos().start,
                ))
            }
            t => parse_error(
                Expected("operator".to_owned(), Some(format!("{:?}", t))),
                self.token_pos().start,
            ),
        }
    }

    // left binding power
    fn lbp(&self) -> u32 {
        match self.token_type() {
            TokenType::LParen => prec(Precedence::Call),
            TokenType::Times | TokenType::Divide => prec(Precedence::Product),
            TokenType::Plus | TokenType::Minus => prec(Precedence::Sum),
            _ => 0,
        }
    }
}

fn token_to_binop(token: Token) -> Result<BinOp, String> {
    match token.token_type() {
        TokenType::Times => Ok(BinOp::Mul),
        TokenType::Divide => Ok(BinOp::Div),
        TokenType::Plus => Ok(BinOp::Add),
        TokenType::Minus => Ok(BinOp::Sub),
        _ => Err("no binop for token".to_owned()),
    }
}

impl<'a> Parser<'a> {
    pub fn new(input: &'a Vec<Token>) -> Parser<'a> {
        Parser {
            input: input.iter().peekable(),
            input_end_pos: input[input.len() - 1].token_pos().end,
        }
    }

    pub fn statement(&mut self) -> ParseResult<Stmt> {
        let expr = self.expression(0)?;
        let pos = expr.pos();

        Ok(Stmt::Expr(expr, pos))
    }

    pub fn expression(&mut self, rbp: u32) -> ParseResult<Expr> {
        let mut left = self.parse_nud()?;
        while self.next_binds_tighter(rbp) {
            left = self.parse_led(left)?;
        }

        Ok(left)
    }

    pub fn program(&mut self) -> ParseResult<Program> {
        let mut decls: Vec<Decl> = vec![];

        while let Some(_) = self.input.peek() {
            let decl = self.decl()?;
            decls.push(decl);
        }

        Ok(Program {
            decls: decls,
            end: self.input_end_pos,
        })
    }

    pub fn decl(&mut self) -> ParseResult<Decl> {
        let shape = self.shape()?;
        Ok(Decl::ShapeDecl(shape))
    }

    pub fn shape(&mut self) -> ParseResult<Shape> {
        // word shape
        let start = self.parse_reserved_word("shape")?;

        // shape name
        let (ident, ident_pos) = self.parse_ident()?;

        // left paren
        if let None = self.match_next(TokenType::LParen) {
            return parse_error(Expected("'(' after shape name".to_owned(), None), ident_pos);
        }

        // args list
        let mut args: Vec<Ident> = vec![];

        if self.next_token_type() != Some(TokenType::RParen) {
            let (arg, _) = self.parse_ident()?;
            args.push(arg);

            while let Some(_) = self.match_next(TokenType::Comma) {
                let (arg, _) = self.parse_ident()?;
                args.push(arg);
            }
        }

        let token = self.input.next();

        // right paren
        match token.map(|t| t.token_type()) {
            Some(TokenType::RParen) => (),
            Some(t) => {
                return parse_error(
                    Expected(
                        "expecting ')' to close function call.".to_owned(),
                        Some(format!("{:?}", t)),
                    ),
                    token.unwrap().token_pos().start,
                )
            }
            None => return parse_error(UnExpectedEndOfInput, self.input_end_pos),
        };

        // block
        let block = self.parse_block()?;

        Ok(Shape {
            name: ident,
            args: args,
            block: block,
            pos: start,
        })
    }

    fn next_binds_tighter(&mut self, rbp: u32) -> bool {
        self.input.peek().map_or(false, |t| t.lbp() > rbp)
    }

    pub fn parse_ident(&mut self) -> ParseResult<(Ident, Pos)> {
        let expr = self.parse_nud()?;
        match &expr {
            Expr::Name(n, _) => {
                if is_reserved_word(&n.as_str()) {
                    parse_error(IdentiferCannotBeReservedWord(n.to_string()), expr.pos())
                } else {
                    Ok((n.clone(), expr.pos()))
                }
            }
            e => parse_error(Expected("identifier".to_owned(), None), e.pos()),
        }
    }

    pub fn parse_reserved_word(&mut self, word: &str) -> ParseResult<Pos> {
        let token = match self.consume() {
            Some(t) => t,
            None => return parse_error(UnExpectedEndOfInput, self.input_end_pos),
        };

        match token.token_type() {
            TokenType::Ident(s) => {
                if s == word {
                    Ok(token.pos())
                } else {
                    parse_error(Expected(word.to_owned(), Some(s)), token.pos())
                }
            }
            _ => parse_error(Expected("identifier".to_owned(), None), token.pos()),
        }
    }

    pub fn parse_function_call(&mut self) -> ParseResult<FunCall> {
        let (ident, ident_pos) = self.parse_ident()?;

        // left paren
        let token = self.input.next();
        match token.map(|t| t.token_type()) {
            Some(TokenType::LParen) => (),
            Some(_) => {
                return parse_error(
                    Expected("'(' to after function name".to_owned(), None),
                    token.unwrap().token_pos().start,
                )
            }
            None => return parse_error(UnExpectedEndOfInput, self.input_end_pos),
        }

        // args
        let mut args: Vec<NamedArg> = vec![];
        if self.next_token_type() != Some(TokenType::RParen) {
            let mut arg = self.parse_named_arg()?;
            args.push(arg);

            while let Some(_) = self.match_next(TokenType::Comma) {
                arg = self.parse_named_arg()?;
                args.push(arg);
            }
        }

        // right paren
        let token = self.input.next();
        match token.map(|t| t.token_type()) {
            Some(TokenType::RParen) => (),
            Some(t) => {
                return parse_error(
                    Expected(
                        "')' to close function call".to_owned(),
                        Some(format!("{:?}", t)),
                    ),
                    token.unwrap().token_pos().start,
                )
            }
            None => return parse_error(UnExpectedEndOfInput, ident_pos),
        }

        Ok(FunCall {
            ident: ident,
            args: args,
            range: create_range(ident_pos, token.unwrap().token_pos().start),
        })
    }

    pub fn parse_block(&mut self) -> ParseResult<Block> {
        // left curly
        let token = self.input.next();
        match token.map(|t| t.token_type()) {
            Some(TokenType::LCurly) => (),
            Some(_) => {
                return parse_error(
                    Expected("'{' to begin block".to_owned(), None),
                    token.unwrap().token_pos().start,
                )
            }
            None => return parse_error(UnExpectedEndOfInput, self.input_end_pos),
        }

        let mut calls: Vec<FunCall> = vec![];

        let mut last = self.match_next(TokenType::RCurly);
        while let None = last {
            let call = self.parse_function_call()?;
            calls.push(call);

            last = self.match_next(TokenType::RCurly);
        }

        match &last {
            Some(_) => Ok(Block {
                calls: calls,
                range: create_range(
                    token.unwrap().token_pos().start,
                    last.unwrap().token_pos().end,
                ),
            }),
            None => parse_error(
                Expected("block to end with '}}'".to_owned(), None),
                token.unwrap().token_pos().start,
            ),
        }
    }

    pub fn parse_named_arg(&mut self) -> ParseResult<NamedArg> {
        let (name, _) = self.parse_ident()?;
        self.match_next(TokenType::Colon);

        let e = match self.expression(0) {
            Err(err) => {
                return parse_error(
                    Expected(
                        "parameters to functions to be in format `(name: value)`".to_owned(),
                        None,
                    ),
                    err.pos(),
                )
            }
            Ok(expr) => expr,
        };

        Ok(NamedArg {
            name: name,
            expr: e,
        })
    }

    pub fn match_next(&mut self, token_type: TokenType) -> Option<Token> {
        let t = match &self.input.peek() {
            Some(t) => **t,
            None => return None,
        };

        if t.token_type() == token_type {
            self.consume();
            Some(t.clone())
        } else {
            None
        }
    }

    pub fn next_token_type(&mut self) -> Option<TokenType> {
        self.input.peek().map(|t| t.token_type())
    }

    fn consume(&mut self) -> Option<Token> {
        self.input.next().map(|t| t.clone())
    }

    fn parse_nud(&mut self) -> ParseResult<Expr> {
        self.input
            .next()
            .map_or(parse_error(UnExpectedEndOfInput, self.input_end_pos), |t| {
                t.nud(self)
            })
    }

    fn parse_led(&mut self, expr: Expr) -> ParseResult<Expr> {
        self.input
            .next()
            .map_or(parse_error(UnExpectedEndOfInput, self.input_end_pos), |t| {
                t.led(self, expr)
            })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot_matches;

    use super::*;
    use crate::lexer;
    use crate::utils::*;

    fn parse_expression(input: &String) -> ParseResult<Expr> {
        let tokens = lexer::lex(input).unwrap();
        let mut parser = Parser::new(&tokens);
        parser.expression(0)
    }

    fn parse_shape(input: &String) -> ParseResult<Shape> {
        let tokens = lexer::lex(input).unwrap();
        let mut parser = Parser::new(&tokens);
        parser.shape()
    }

    #[test]
    fn parse_number_literal() {
        let ast = parse_expression(&"2".to_owned());
        assert_debug_snapshot_matches!(ast);
    }

    #[test]
    fn parse_simple_identifier() {
        let ast = parse_expression(&"hello".to_owned());
        assert_debug_snapshot_matches!(ast);
    }

    #[test]
    fn parse_reserved_identifier() {
        let ast = parse_expression(&"shape".to_owned());

        match ast {
            Ok(_) => assert!(false, "reserved word should not be parsable as identifier"),
            Err(_) => assert!(true),
        }
    }

    #[test]
    fn parse_simple_binary_expression() {
        let ast = parse_expression(&"10 + 2".to_owned());
        assert_debug_snapshot_matches!(ast);
    }

    #[test]
    fn parse_simple_shape() {
        let code = "shape circle(r) {
  ellipse(rx: r, ry: r)
}";

        let ast = parse_shape(&code.to_owned());
        assert_debug_snapshot_matches!(ast);
    }

    #[test]
    fn parse_shape_multiple_statements() {
        let code = "shape circle(r) {
  ellipse(rx: r, ry: r * 10)
  ellipse(rx: r, ry: r)
  ellipse(rx: r, ry: r * 10)
  ellipse(rx: r, ry: r * 10)
  ellipse(rx: r, ry: r)
  ellipse(rx: r, ry: r)
}";

        let ast = parse_shape(&code.to_owned());
        assert_debug_snapshot_matches!(ast);
    }

    #[test]
    fn parse_simple_program() {
        let code = "
shape shape1(r) {
  circle(r: 10)
}

shape shape2(r) {
  circle(r: 20)
}
";

        let tokens = lexer::lex(&code.to_owned()).unwrap();
        let program = parse_program(tokens).unwrap();

        assert_eq!(2, program.decls.len());
        assert_debug_snapshot_matches!(program);
    }
}
