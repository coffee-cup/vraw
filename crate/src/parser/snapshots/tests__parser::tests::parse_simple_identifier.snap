---
created: "2019-07-08T18:41:37.171274Z"
creator: insta@0.8.1
source: src/parser/mod.rs
expression: ast
---
Ok(
    Name(
        "hello",
        Range {
            start: Pos {
                line: 0,
                column: 0,
            },
            end: Pos {
                line: 0,
                column: 5,
            },
        },
    ),
)
